app/questionnaire/navigator.py

```diff
@@ -0,0 +1,182 @@
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def evaluate_rule(rule, answer):
+    """
+    Determine whether a rule will be satisfied based on a given answer
+    :param rule:
+    :param answer:
+    :return:
+    """
+    when = rule['when']
+    match_value = when['value']
+    condition = when['condition']
+
+    # Evaluate the condition on the routing rule
+    if condition == 'equals' and match_value == answer:
+        return True
+    elif condition == 'not equals' and match_value != answer:
+        return True
+    return False
+
+
+def visit_block(blocks, block_id, answers, path):
+    """
+    Recursive method which visits all the blocks and returns path taken
+    given a list of answers
+
+    :param blocks: A list containing all blocks in the survey
+    :param block_id: The block id to visit
+    :param answers: The answers to use on evaluating rules
+    :param path: The known path as a list which has been visited already
+    :return: A list of block ids followed through the survey
+    """
+    if block_id in Navigator.CLOSING_INTERSTITIAL_PATH:
+        return path
+
+    path.append(block_id)
+
+    # Return the index of the block id to be visited
+    block_id_index = next(index for (index, b) in enumerate(blocks) if b["id"] == block_id)
+
+    if 'routing_rules' in blocks[block_id_index]:
+        for rule in blocks[block_id_index]['routing_rules']:
+            goto_rule = rule['goto']
+            if 'when' in goto_rule.keys():
+                answer_index = goto_rule['when']['id']
+                if answer_index in answers:
+                    answer = answers[answer_index]
+                    if evaluate_rule(goto_rule, answer):
+                        return visit_block(blocks, goto_rule['id'], answers, path)
+                else:
+                    return path
+            elif 'id' in goto_rule.keys():
+                return visit_block(blocks, goto_rule['id'], answers, path)
+    else:
+        if block_id_index != len(blocks) - 1:
+            next_block_id = blocks[block_id_index + 1]['id']
+            return visit_block(blocks, next_block_id, answers, path)
+    return path
+
+
+class Navigator:
+    PROCEEDING_INTERSTITIAL_PATH = ['introduction']
+    CLOSING_INTERSTITIAL_PATH = ['summary', 'thank-you']
+
+    def __init__(self, survey_json):
+        self.survey_json = survey_json
+        self.blocks = self.get_blocks()
+        self.first_block_id = self.blocks[0]['id']
+
+    def get_routing_path(self, answers=None):
+        """
+        Returns a list of the block ids visited based on answers provided
+        :return: List of block ids
+        """
+        answers = answers or {}
+        routing_path = []
+
+        return visit_block(self.blocks, self.first_block_id, answers, routing_path)
+
+    def get_location_path(self, answers=None):
```

Comment:
I'm not sure, given the list of locations haven't yet been visited, at least from a respondents point of view.